### ISSUE TITLE
Phase 6.3 — Wire EWCRegulariser into STDPOnlineLearner: penalty accumulation and LEARNING-boundary consolidation

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/asi-build#249]
## Overview

Phase 6.3 integrates the `EWCRegulariser` (Phase 6.1, #241) with the `STDPOnlineLearner` (#181) to enforce **continual-learning constraints** during the LEARNING phase of the `CognitiveCycle`. This is the step that makes EWC _active_: every weight update is penalised proportional to Fisher importance before it is applied.

Companion discussions: #250 (penalty math + visualization), #251 (tuning & debugging Q&A)  
Previous phase: #245 (Fisher backends — Neo4j + Redis + FisherStoreFactory)

---

## Background

`STDPOnlineLearner` currently applies raw STDP weight deltas:

```python
delta_w = A_plus * exp(-dt / tau_plus)   # LTP
        - A_minus * exp(dt  / tau_minus)  # LTD
w += learning_rate * delta_w
```

After Phase 6.3, the update becomes:

```python
ewc_penalty = ewc_lambda * sum(F_i * (w_i - w_star_i)**2)
w += learning_rate * (delta_w - grad(ewc_penalty))
```

where `F_i` is the Fisher importance for parameter `i` and `w_star_i` is the anchor weight after the previous SLEEP_PHASE consolidation.

---

### 6.3.1  `EWCRegulariser.penalty_gradient(weights)` (new method)

Returns per-weight gradient of the EWC quadratic penalty:

```python
async def penalty_gradient(
    self,
    weights: npt.NDArray[np.float32],
    task_id: str,
) -> npt.NDArray[np.float32]:
    """Return 2 * lambda * F * (w - w_star)."""
    snapshot = await self._store.load(task_id)
    if snapshot is None:
        return np.zeros_like(weights)
    grad = (
        2.0
        * self._config.ewc_lambda
        * snapshot.fisher_diag
        * (weights - snapshot.anchor_weights)
    )
    return np.clip(grad, -self._config.max_penalty_clip, self._config.max_penalty_clip)
```

### 6.3.2  `STDPOnlineLearner._apply_ewc_constraint(delta_w, task_id)` (new method)

```python
async def _apply_ewc_constraint(
    self,
    delta_w: npt.NDArray[np.float32],
    task_id: str,
) -> npt.NDArray[np.float32]:
    if self._ewc is None:
        return delta_w   # EWC not configured — pass-through
    grad = await self._ewc.penalty_gradient(self._weights, task_id)
    return delta_w - grad
```

### 6.3.3  `STDPOnlineLearner.update()` — modified signature

```python
async def update(
    self,
    pre_times:  npt.NDArray[np.float32],
    post_times: npt.NDArray[np.float32],
    task_id:    str = "default",
) -> WeightUpdate:
    delta_w = self._stdp_kernel(pre_times, post_times)
    delta_w = await self._apply_ewc_constraint(delta_w, task_id)
    self._weights = np.clip(
        self._weights + self._config.learning_rate * delta_w,
        self._config.w_min,
        self._config.w_max,
    )
    return WeightUpdate(delta_w=delta_w, new_weights=self._weights.copy())
```

### 6.3.4  `EWCConfig` — new fields

| Field | Type | Default | Purpose |
|---|---|---|---|
| `max_penalty_clip` | `float` | `1.0` | Gradient clip for numerical stability |
| `penalty_accumulation` | `Literal["per_update", "batch"]` | `"per_update"` | When to apply penalty |

### 6.3.5  Prometheus metrics (new)

| Metric | Type | Labels | Meaning |
|---|---|---|---|
| `phase6_ewc_penalty_magnitude` | Histogram | `task_id` | L2 norm of penalty gradient per update |
| `phase6_ewc_penalty_clipped_total` | Counter | `task_id` | Clip events (stability guard) |
| `phase6_weight_delta_before_ewc` | Histogram | — | Raw STDP delta L2 before EWC |
| `phase6_weight_delta_after_ewc` | Histogram | — | Delta L2 after EWC constraint |

---

## Acceptance Criteria

- [ ] `EWCRegulariser.penalty_gradient()` returns zeros when no snapshot exists
- [ ] `STDPOnlineLearner` accepts optional `ewc: EWCRegulariser | None` in constructor
- [ ] Weight updates respect EWC constraint when `ewc` is configured
- [ ] `max_penalty_clip` prevents gradient explosion (unit test: Fisher=1e6)
- [ ] Pass-through works correctly when `ewc=None` (backward compat)
- [ ] Prometheus metrics emit on every `update()` call
- [ ] mypy strict passes (no `Any` escapes in penalty path)
- [ ] Existing STDP tests still pass with `ewc=None` default

---

## Test Targets (10)

| Test | What it verifies |
|---|---|
| `test_penalty_gradient_no_snapshot` | Returns zeros when Fisher store empty |
| `test_penalty_gradient_values` | Correct 2λF(w-w*) formula |
| `test_penalty_gradient_clip` | Clips to `max_penalty_clip` when Fisher is huge |
| `test_update_without_ewc` | Pass-through: `ewc=None` → same as pre-Phase-6 |
| `test_update_with_ewc_reduces_delta` | EWC penalty reduces effective learning rate |
| `test_update_ewc_zero_fisher` | Fisher=0 → gradient=0 → delta unchanged |
| `test_ewc_anchor_respected` | Weights near anchor → small penalty; far → large |
| `test_prometheus_penalty_histogram` | `phase6_ewc_penalty_magnitude` emits on update |
| `test_prometheus_clip_counter` | Counter increments on clip event |
| `test_mypy_strict` | No type errors in integration path |

---

## Implementation Order

1. Add `max_penalty_clip` + `penalty_accumulation` to `EWCConfig` (learning/ewc.py)
2. Implement `EWCRegulariser.penalty_gradient()` with Prometheus hooks
3. Add `ewc: EWCRegulariser | None = None` to `STDPOnlineLearner.__init__()`
4. Implement `_apply_ewc_constraint()` method
5. Modify `update()` to call constraint method (backward-compat: `task_id="default"`)
6. Write 10 unit tests (mock Fisher store — no real Neo4j/Redis)
7. Run `mypy --strict asi/phase6/`
8. Update `build_phase5_cycle()` factory to accept optional `EWCRegulariser`

---

## References
- #241 — Phase 6.1 EWC foundation
- #245 — Phase 6.2 Fisher backends
- #181 — STDPOnlineLearner
- Discussion #248 — Phase 6.3 integration strategy (this issue implements Option B)
- Wiki: [Phase-6-EWC-Foundation](https://github.com/web3guru888/asi-build/wiki/Phase-6-EWC-Foundation)
- Wiki: [Phase-6-Fisher-Backends](https://github.com/web3guru888/asi-build/wiki/Phase-6-Fisher-Backends)

**Comment by web3guru888** [comment]
## Implementation Notes — Phase 6.3

### Module placement

```
asi/
  phase6/
    learning/
      ewc.py           ← add penalty_gradient() here (EWCRegulariser)
      fisher.py        ← add max_penalty_clip to EWCConfig here
    stdp/
      learner.py       ← modify STDPOnlineLearner here
tests/
  phase6/
    test_ewc_integration.py    ← 10 new tests
```

### Key implementation notes

**1. `penalty_gradient()` must handle mismatched array shapes**

If `anchor_weights.shape != weights.shape` (weight matrix resized between tasks), return zeros and log a warning — don't raise. Fisher snapshots from an older topology should be silently discarded:

```python
if snapshot.anchor_weights.shape != weights.shape:
    logger.warning(
        "Fisher snapshot shape mismatch: %s vs %s — skipping EWC",
        snapshot.anchor_weights.shape, weights.shape
    )
    return np.zeros_like(weights)
```

**2. Async Fisher store — don't await inside a hot loop**

`penalty_gradient()` is `async` and awaits `self._store.load()`. For `InMemoryFisherStore` this is essentially free (asyncio.Lock, no I/O). For `Neo4jFisherStore`, it's a network round-trip — not acceptable per-update.

Recommended pattern: **cache the snapshot in `STDPOnlineLearner`** and refresh only on task switch or SLEEP boundary:

```python
async def _refresh_fisher_cache(self, task_id: str) -> None:
    self._cached_snapshot = await self._ewc._store.load(task_id)
    self._cached_task_id = task_id
```

This makes per-update cost O(1) array ops regardless of backend.

**3. The `task_id` default value**

Using `task_id="default"` as the default in `update()` means single-task agents work without any configuration change. Multi-task agents must pass explicit task IDs. Document this clearly.

**4. mypy strict — the `None` check pattern**

```python
# Wrong (mypy error if ewc is Optional[EWCRegulariser]):
if self._ewc:
    grad = await self._ewc.penalty_gradient(...)  # mypy ok
    
# Also correct — type narrowing via isinstance:
if isinstance(self._ewc, EWCRegulariser):
    ...
```

**5. Test isolation — mock the Fisher store, not the whole EWCRegulariser**

For unit tests, mock `FisherMatrixBase` (the ABC), not `EWCRegulariser` itself — we want to test the real gradient computation:

```python
@pytest.fixture
def mock_store(snapshot: FisherSnapshot | None):
    store = AsyncMock(spec=FisherMatrixBase)
    store.load.return_value = snapshot
    return store

@pytest.fixture
def regulariser(mock_store):
    config = EWCConfig(ewc_lambda=100.0, max_penalty_clip=1.0)
    return EWCRegulariser(store=mock_store, config=config)
```

**6. Prometheus pre-init (avoid LookupError on first scrape)**

Pre-initialize all label combinations at module load:

```python
# In phase6_metrics.py or ewc.py module init:
for task_id in ("default",):
    PENALTY_MAGNITUDE.labels(task_id=task_id)  # pre-init histogram
    PENALTY_CLIPPED.labels(task_id=task_id)    # pre-init counter
```

See Discussion #223 for the full Prometheus pitfalls guide.

---

Related: Discussion #250 (penalty math + visualization) · Discussion #251 (tuning Q&A)
